### ISSUE TITLE
LaTeX export: sphinxhref cannot handle some unicode characters in name of hyperlink

### ISSUE THREAD

**Issue by malparti** [sphinx-doc/sphinx#9321]
`\sphinxhref` cannot handle some unicode characters, such as ș, in the name of hyperlinks.

**To Reproduce**
Define a link as follows:
```
`Brașov <https://www.unitbv.ro/>`_
```
When compiling to LaTeX, this is translated as
```latex
\sphinxhref{https://www.unitbv.ro/}{Brașov}
```
but the ș does not work (at least with the versions of pdflatex I have tried), even though LaTeX is properly configured to deal with ș. In particular,
```latex
\href{https://www.unitbv.ro/}{Brașov}
```
works fine. See the following screen shot for a comparison of the  what the two commands yield:

![2021-06-11_19-08](https://user-images.githubusercontent.com/85755504/121716389-a38bbe80-cae8-11eb-8a03-c25230124d91.png)

**Environment info**
 Version of pdflatex:
```
pdfTeX 3.14159265-2.6-1.40.19 (TeX Live 2019/dev/Debian)
kpathsea version 6.3.1/dev
Copyright 2018 Han The Thanh (pdfTeX) et al.
There is NO warranty.  Redistribution of this software is
covered by the terms of both the pdfTeX copyright and
the Lesser GNU General Public License.
For more information about these matters, see the file
named COPYING and the pdfTeX source.
Primary author of pdfTeX: Han The Thanh (pdfTeX) et al.
Compiled with libpng 1.6.34; using libpng 1.6.36
Compiled with zlib 1.2.11; using zlib 1.2.11
Compiled with xpdf version 4.00
```

I am using Sphinx 4.0.2, which has the following definition for `\sphinxhref`:
```tex
% Protect \href's first argument in contexts such as sphinxalltt (or
% \sphinxcode). Sphinx uses \#, \%, \& ... always inside \sphinxhref.
\protected\def\sphinxhref#1#2{{%
    \sphinxunactivateextrasandspace % never do \scantokens with active space!
% for the \endlinechar business, https://github.com/latex3/latex2e/issues/286
    \endlinechar\m@ne\everyeof{{\endlinechar13 #2}}% keep catcode regime for #2
    \scantokens{\href{#1}}% normalise it for #1 during \href expansion
}}
```

**Comment by malparti** [comment]
This problem only occurs with utf8x, i.e. when
```latex
\usepackage[utf8x]{inputenc}
```
is used. When using
```latex
\usepackage[utf8]{inputenc}
```
there is not problem.

I had switched from utf8 to utf8x in an attempt to fix another problem, but I'm happy to go back to utf8, which is what I normally use anyway. So, even though I still think this is a bug of \sphinxhref (since it works with \href but not with \sphinxhref), my particular problem is solved. I'm happy to do some more testing if needed.

**Comment by jfbu** [comment]
Weird indeed but here is the explanation: with `utf8x` the following procedure applies (taken from `ucs` documentation, `texdoc ucs`)

```
1.5 Known problems
Note, that if a character from some not yet loaded Unicode page appears, a file has to be loaded.
```

However the `\sphinxhref` uses `\scantokens` in a way requiring `\everyeof` usage. As the name indicates this inserts tokens at end of every loaded file. Normally this is meant to act only on the pseudo-file used internally to TeX by `\scantokens`.  But here when `ș` is encountered the first time, a real external file will be loaded.

To be precise it is 

```
/usr/local/texlive/2021/texmf-dist/tex/latex/ucs/data/uni-2.def
File: uni-2.def 2013/05/13 UCS: Unicode data U+0200..U+02FF
```

Because the Unicode code point of `ș` is `U+0219` hence the `2` in `uni-2.def`. You can fix the problem by adding `\PreloadUnicodePage{2}` to the document preamble.

Or `\PrerenderUnicode{ș}` is also possible.

A fix at Sphinx level would be for `\sphinxhref` to systematically do `\PrerenderUnicode{the second argument}` in case `utf8x` is loaded.

I hesitate between adding this fix, or rather adding to documentation advice to use `\PreloadUnicodePage{suitable number as per ucs documentation}` in the preamble.

Indeed LaTeX team is maintaining `utf8`,  not `utf8x`, and those last few years, LaTeX team has added quite a few things in this area; I am not sure if `utf8x` is compatible with for example recent LaTeX's `\input{filename}` being compatible with filename having Unicode characters. Not tested. Well now I have tested and it breaks :

```latex
\documentclass{article}
%\usepackage{ucs}
\usepackage[utf8x]{inputenc}
\usepackage[T1]{fontenc}

%\PreloadUnicodePage{2}
%\PrerenderUnicode{ș}
\begin{document}

\input{Brașov}
\end{document}
```

where a file `Brașov.tex` exists in same repertory. There is no issue with `utf8` (which since LaTeX 2018 I think is default and one can drop the `inputenc`). Try the above with the `\usepackage[utf8x]{inputenc}` commented out, it works fine with a recent LaTeX.

If LaTeX team does not make a "fix" to `\input` in presence of `utf8x`, I am not much motivated to make one to `\sphinxhref`.

Try a Sphinx project with an image named `Brașov.png` and it will break if `utf8x` option is used except if letter `ș` has already been encountered in the document.

Confirmation:
```
! Missing endcsname inserted.
<to be read again> 
                   begingroup 
l.84 ...ndent\sphinxincludegraphics{{Brașov}.png}
                                                  
? 
```

from this index.rst file:

```rst
Welcome to FOO's documentation!
===============================

.. image:: Brașov.png
```

and

```
latex_elements = {'inputenc': "\\usepackage[utf8x]{inputenc}"}
```

in conf.py.

<strike>But the problem shows only if `ș` has not been encountered yet in a safer place in the document.</strike> The same test file works perfectly fine with `utf8`.

EDITED: with `utf8x` the LaTeX build is broken from image inclusion with filename `Brașov.png`,  even if letter `ș` has already been encountered. The `\PreloadUnicodePage{2}` trick does not work here.

In conclusion and in view of the above, LaTeX is not fully compatible with `utf8x` in a way which goes beyond Sphinx specifics.  Users of `utf8x` should take the extra steps like using `\PreloadUnicodePage{<N>}`.

Having thought about this there will be no "fix" on Sphinx side. Perhaps some addition to documentation but it would only be advice to read `utf8x` (aka `ucs`)  documentation.

**Comment by jfbu** [comment]
Post scriptum regarding `\input` vs `utf8x`

```
\documentclass{article}
%\usepackage{ucs}
\usepackage[utf8x]{inputenc}
\usepackage[T1]{fontenc}

\PreloadUnicodePage{2}% DOES NOT HELP
%\PrerenderUnicode{ș}
\begin{document}

\input{Brașov}
\end{document}
```

In fact as indicated above, the `\input` with filename containing Unicode characters is broken with `utf8x` and the `\PreloadUnicodePage` trick **does not work**.

Certainly image inclusion with such Unicode filenames is broken too, if `utf8x` is present.

In short, LaTeX more and more assumes Unicode support is from `utf8` and not `utf8x`.  So using `utf8x` induces some limitations.

However, it still works with the `\PreloadUnicodePage` suitably added to preamble in your context.

By the way thanks for your report and the pre-analysis that it was `utf8x`!

**Comment by malparti** [comment]
Thanks for the detailed reply.

That also explains a puzzling behavior I had observed: after opening this issue, I tried to change a few things here and there in the .tex file to see if I could understand what was going on. At some point, I was no longer able to reproduce the issue and I thought it was due to the fact that had used a `\newline` before `\sphinxhref`, or something like that... But nothing really made sense. Now I understand that it was simply because I had (unknowingly) put a "ș" earlier in the document.

After a bit of googling I had found the `\PrerenderUnicode{ș}` work-around, but in my case using utf8 is the best solution.

I think mentioning this somewhere in the documentation would indeed be useful. It would have saved me a bit of time, and since there are a lot of easy "suboptimal" work-arounds (such as using "s" instead of "ș", or using `.. raw:: html` + `.. raw:: latex`, which was what I was going to settle for before I found out everything works fine with utf8), I suspect I may not be the first person to have ran into this problem.

**Comment by jfbu** [comment]
Thanks for feedback. What do you think of #9332?

**Comment by malparti** [comment]
@jfbu Sorry I forgot to reply but I think #9332 is good and that the issue can probably be closed!

Thanks!

**Comment by jfbu** [comment]
@malparti I have merged the documentation update.  Certainly users can get bitten by the issue you encountered, even unknowingly if they generate PDF as an aside to say HTML and do not check output.  Already Sphinx has some extras to make it work with utf8x but as explained upstream LaTeX utf8 has per default more coverage of Unicode it had ten years ago (and it keeps being extended), the support of Unicode by pdflatex has evolved regarding filenames in particular,  but this is utf8x incompatible.
